rolling-shutter/p2p/floodsubpeerdiscovery/gossippeerdiscovery.go

```diff
@@ -0,0 +1,113 @@
+package floodsubpeerdiscovery
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	pubsub "github.com/libp2p/go-libp2p-pubsub"
+	"github.com/libp2p/go-libp2p/core/peerstore"
+	"github.com/rs/zerolog/log"
+	"github.com/shutter-network/rolling-shutter/rolling-shutter/medley/encodeable/address"
+	"google.golang.org/protobuf/proto"
+)
+
+const defaultTopic = "_peer-discovery._p2p._pubsub"
+
+type FloodsubPeerDiscovery struct {
+	PeerDiscoveryComponents
+	Interval int
+	Topics   []*pubsub.Topic
+}
+
+type PeerDiscoveryComponents struct {
+	PeerId    address.P2PIdentifier
+	Pubsub    *pubsub.PubSub
+	PeerStore peerstore.Peerstore
+}
+
+func (pd *FloodsubPeerDiscovery) Init(config PeerDiscoveryComponents, interval int, topics []string) error {
+	pd.Interval = interval
+	pd.PeerId = config.PeerId
+	pd.Pubsub = config.Pubsub
+	pd.PeerStore = config.PeerStore
+
+	if len(topics) > 0 {
+		for _, topic := range topics {
+			topic, err := pd.Pubsub.Join(topic)
+			if err != nil {
+				return fmt.Errorf("failed to join topic | err %v", err)
+			}
+			pd.Topics = append(pd.Topics, topic)
+		}
+	} else {
+		topic, err := pd.Pubsub.Join(defaultTopic)
+		if err != nil {
+			return fmt.Errorf("failed to join topic | err %v", err)
+		}
+		pd.Topics = append(pd.Topics, topic)
+	}
+	return nil
+}
+
+func (pd *FloodsubPeerDiscovery) Start(ctx context.Context) error {
+	timer := time.NewTicker(time.Duration(pd.Interval) * time.Second)
+
+	for {
+		select {
+		case <-timer.C:
+			err := pd.broadcast()
+			if err != nil {
+				log.Info().Msgf("error in broadcasting floodsub msg | %v", err)
+				return err
+			}
+		case <-ctx.Done():
+			return nil
+		}
+	}
+}
+
+func (pd *FloodsubPeerDiscovery) broadcast() error {
+	pubKey, err := pd.PeerId.ExtractPublicKey()
+	if err != nil {
+		return fmt.Errorf("peerId was missing public key | err %v", err)
```

Comment:
With `fmt.Errorf` using `%w` instead of `%v` for errors makes more sense since it wraps the error and allows higher up code to inspect the underlying cause.